### PR TITLE
Initial support for module: node12

### PR DESCRIPTION
**Pull request by weswigham**
This PR adds the basic framework for (compiler options and defaults) and core emit behavior for a new `module: node12` and `module: nodenext` compiler option. Both should be almost identical, for now.

The core of this PR is allowing `.js` and `.ts` files to be interpreted as either esm or cjs format files, within the same compilation.
* The esm emit branch is identical to es2020+ module emit, except we transpile TS-style `import = require` statements down to a helper-like `createRequire` call, rather than error'ing and eliding them. (This gives easy access to sync `require` statements as-needed for, eg, conditional top-level access to cjs dependencies)
* The cjs emit branch modifies normal cjs emit to preserve `import` call expressions as-is, this way they can be used to pull in async (esm) dependencies within cjs-targeted files, and is otherwise normal cjs output.
* Since we don't currently actually read in non-ts and non-js extensions (yet - that'll be a followup pull request), the mode of the `.ts` or `.js` file is determined by the enclosing `package.json` and its `type` field (if it is present and set to `module`, `.ts` and `.js` files are esm format, otherwise they are cjs format).

As far as checking goes, both options imply `esModuleInterop` for now (this only really affects default import checking and cjs module emit and may be subject to change when more checker behaviors are added later if they pickup the work this flag does). Additionally, `module: node12` implies a `moduleResolution: node12` and `target: es2020`, and `module: nodenext` implies a `moduleResolution: nodenext` and `target: esnext`. The new `moduleResolution` modes currently just redirect to the old `node` resolution mode until they're actually implemented in another PR.

Also included is a driveby fix for js declaration emit where `export {default} from "fs";` in js was mistakenly serialized to `export {"fs" as default} from "fs";` in a declaration file when `"fs"` is just a shorthand module declaration.

### CODE REVIEW

**Comment by weswigham** [comment]
@typescript-bot pack this

**Comment by weswigham** [comment]
@typescript-bot pack this

**Comment by weswigham** [comment]
@typescript-bot test this

**Comment by weswigham** [comment]
@typescript-bot test this

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -7141,8 +7168,10 @@ namespace ts {
                     if (!target) {
                         return;
                     }
-                    let verbatimTargetName = unescapeLeadingUnderscores(target.escapedName);
-                    if (verbatimTargetName === InternalSymbolName.ExportEquals && (compilerOptions.esModuleInterop || compilerOptions.allowSyntheticDefaultImports)) {
+                    // If `target` refers to a shorthand module symbol, the name we're trying to pull out isn;t recoverable from the target symbol
```
Comment (andrewbranch) [inline src/compiler/checker.ts:7171]: This is the drive-by fix, right? What’s a shorthand module symbol?

```suggestion
                    // If `target` refers to a shorthand module symbol, the name we're trying to pull out isn't recoverable from the target symbol
```

**Review by andrewbranch — CHANGES_REQUESTED** [review]
A couple questions; the comment in esnextAnd2015.ts is the only one blocking for me.

File: src/compiler/transformers/module/esnextAnd2015.ts [hunk 2]
```diff
@@ -55,8 +69,10 @@ namespace ts {
         function visitor(node: Node): VisitResult<Node> {
             switch (node.kind) {
                 case SyntaxKind.ImportEqualsDeclaration:
-                    // Elide `import=` as it is not legal with --module ES6
-                    return undefined;
+                    // Though an error in es2020 modules, in node-flavor es2020 modules, we can helpfully transform this to a synthetic `require` call
+                    // To give easy access to a synchronous `require` in node-flavor esm. We do the transform even in scenarios where we error, but `import.meta.url`
+                    // is available, just because the output is reasonable for a node-like runtime.
+                    return getEmitScriptTarget(compilerOptions) >= ModuleKind.ES2020 ? visitImportEqualsDeclaration(node as ImportEqualsDeclaration) : undefined;
```
[the remark this answers is not in the document]
Comment (weswigham) [reply src/compiler/transformers/module/esnextAnd2015.ts:75]: Well, we'll issue an error on this import form in non-node es2020+, we're just choosing to emit the node-style transpilation, rather than omit the erroneous node, since the node-style transpilation is still pretty reasonable (and could actually work on other runtimes if they're providing a cjs interop layer).

File: src/compiler/types.ts [hunk 3]
```diff
@@ -6099,7 +6119,11 @@ namespace ts {
         //       module kind).
         ES2015 = 5,
         ES2020 = 6,
```
[the remark this answers is not in the document]
Comment (weswigham) [reply src/compiler/types.ts:6121]: Yeah. We don't have a module resolution mode that understands URLs at all, so classic mode is as close as we get.

File: src/compiler/checker.ts (hunk 1 above)
Comment (weswigham) [reply src/compiler/checker.ts:7171]: It's a (member of a) `declare module "foo";` defined module.

**Review by andrewbranch — APPROVED** [review]

File: src/compiler/transformers/module/esnextAnd2015.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (andrewbranch) [reply src/compiler/transformers/module/esnextAnd2015.ts:75]: Got it. I really like the `createRequire` helper, by the way. That’s kind of a pain in vanilla JS.

File: src/compiler/types.ts [hunk 4]
```diff
@@ -3532,6 +3532,20 @@ namespace ts {
...
+         * source file in question is an ESNext-output-format file, or a CommonJS-output-format
+         * module. This is derived by the module resolver as it looks up the file, since
+         * it is derived from either the file extension of the module, or the containing
+         * `package.json` context, and affects both checking and emit.
+         *
+         * It is _public_ so that (pre)transformers can set this field,
+         * since it switches the builtin `node` module transform. Generally speaking, if unset,
+         * the field is treated as though it is `ModuleKind.CommonJS`.
+         */
+        impliedNodeFormat?: ModuleKind.ESNext | ModuleKind.CommonJS;
```
Comment (sandersn) [inline src/compiler/types.ts:3547]: is there a `nodeFormat` property already? Would it be possible to call this `nodeFormat`? Does that name imply more certainty than this one?

File: src/compiler/types.ts [hunk 5]
```diff
@@ -5894,7 +5908,13 @@ namespace ts {
 
     export enum ModuleResolutionKind {
         Classic  = 1,
-        NodeJs   = 2
+        NodeJs   = 2,
+        // Starting with node12, node's module resolver has significant departures from tranditional cjs resolution
```
Comment (sandersn) [inline src/compiler/types.ts:5912]: Does this comment show up in the quick info of our API? If not, I think we could do with a shorter comment for future maintainers.

File: src/compiler/types.ts [hunk 6]
```diff
@@ -5894,7 +5908,13 @@ namespace ts {
 
     export enum ModuleResolutionKind {
         Classic  = 1,
-        NodeJs   = 2
+        NodeJs   = 2,
+        // Starting with node12, node's module resolver has significant departures from tranditional cjs resolution
+        // to better support ecmascript modules and their use within node - more features are still being added, so
+        // we can expect it co change over time, and as such, offer both a `NodeNext` moving resolution target, and a `Node12`
```
Comment (sandersn) [inline src/compiler/types.ts:5914]:
```suggestion
        // we can expect it to change over time, and as such, offer both a `NodeNext` moving resolution target, and a `Node12`

```

File: src/compiler/transformers/module/esnextAnd2015.ts [hunk 7]
```diff
@@ -67,6 +83,106 @@ namespace ts {
             return node;
         }
 
+        /**
+         * Creates a `require()` call to import an external module.
+         *
+         * @param importNode The declararation to import.
```
Comment (sandersn) [inline src/compiler/transformers/module/esnextAnd2015.ts:89]:
```suggestion
         * @param importNode The declaration to import.

```

**Review by sandersn — COMMENTED** [review]
I don't have any useful comments, just a lot of typos. :)

**Review by andrewbranch — CHANGES_REQUESTED** [review]
I just remembered—everything that looks at `UsersPreferences["importModuleSpecifierEnding"]` needs to completely ignore that setting and pretend it’s always `"full"` when `--moduleResolution` is `node12` or `nodenext`. That setting is only applicable in modes where `foo`, `foo/index`, and `foo/index.js` are plausibly interchangeable. As it is now, I believe auto imports and string completions will default to leaving off extensions and `index` filenames, which won’t resolve in node-style module resolution.

**Comment by weswigham** [comment]
@andrewbranch considering this doesn't add _any_ of the new resolution rules yet (except format detection), I think that's probably fine to leave as-is until those are in place.

**Review by andrewbranch — APPROVED** [review]

**Comment by AlbertMarashi** [comment]
What are we waiting on?

**Comment by DanielRosenwasser** [comment]
@AlbertMarashi This is a fairly large change which we will likely land in 4.5. It likely needs further review and we will need to walk it through a few scenarios to see how well it works.

**Comment by weswigham** [comment]
Moreover, we want this PR (the emit mode) to land at the same time as my next PR (the new extension support) and the follow-up PR (the new module resolution behavior), so we don't have partial support for a release. It's just too big to review in a single PR.

File: src/compiler/types.ts (hunk 4 above)
Comment (weswigham) [reply src/compiler/types.ts:3562]: no.... `nodeFormat` just sounds like it might have something to do with `ts.Node` (since it's a property on one), while the extra word `implied` I'm hoping makes it more at-a-glance apparent that it has something to do with node-the-platform.

File: src/compiler/types.ts (hunk 5 above)
Comment (weswigham) [reply src/compiler/types.ts:5932]: No, it doesn't show up in our public API - it's more of a rationale comment than anything else.

**Comment by weswigham** [comment]
Followup PR 1 (extension support) is up here and is targeted against this PR for ease of review: https://github.com/weswigham/TypeScript/pull/67

**Comment by CraigMacomber** [comment]
I tested out the build posted here. It worked flawlessly for the package I had a problem with: just switched to this version of typescript, and my problem was fixed (no tsconfig changes needed: defaults simply fixed it).

I did however have issues with other code when I build our whole mono-repo. Looks like `import type` statements have issues. For example, several of the d.ts file from fluidframework failed to build due to claimed syntax errors in the `imports type` statements.

One example:

> [tsc] ../../node_modules/@fluidframework/routerlicious-driver/dist/documentStorageService.d.ts:5:13 - error TS1005: '=' expected.
> [tsc]
> [tsc] 5 import type { ITelemetryLogger } from "@fluidframework/common-definitions";
> [tsc]               ~
> [tsc]
> [tsc] ../../node_modules/@fluidframework/routerlicious-driver/dist/documentStorageService.d.ts:5:39 - error TS1005: ';' expected.
> [tsc]
> [tsc] 5 import type { ITelemetryLogger } from "@fluidframework/common-definitions";
> [tsc]                                         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> [tsc]

This codebase builds cleanly with typescript `4.3.5` as well as `4.4.0-dev.20210727`, but gets these errors on `npm:@typescript-deploys/pr-build@4.4.0-pr-44501-7`. Since these errors are in imports, I think it might be an issue with this PR.

**Comment by Jack-Works** [comment]
Can I bike shed the name? `node12` => `node-12` and `nodenext` => `node-next`

File: src/compiler/checker.ts [hunk 8]
```diff
@@ -2851,18 +2851,18 @@ namespace ts {
 
         function reportInvalidImportEqualsExportMember(node: ImportDeclaration | ExportDeclaration | VariableDeclaration, name: Identifier, declarationName: string, moduleName: string) {
             if (moduleKind >= ModuleKind.ES2015) {
-                const message = compilerOptions.esModuleInterop ? Diagnostics._0_can_only_be_imported_by_using_a_default_import :
+                const message = getESModuleInterop(compilerOptions) ? Diagnostics._0_can_only_be_imported_by_using_a_default_import :
```
Comment (Jack-Works) [inline src/compiler/checker.ts:2854]: Should we have a normalized version of CompilerOptions so that we don't have to call those function again and again?

**Comment by weswigham** [comment]
`-` has weird command line parsing implications, so we should probably avoid including it in any options.

File: src/compiler/checker.ts (hunk 8 above)
Comment (weswigham) [reply src/compiler/checker.ts:2854]: Eh, it might be nice, but then we'd have to track it through the typesystem and delicate where we have processed vs unprocessed compiler options.

**Comment by Jack-Works** [comment]
> `-` has weird command line parsing implications, so we should probably avoid including it in any options.

Even in form of `--module node-next`? It's surprising

**Comment by Jamesernator** [comment]
> Can I bike shed the name? `node12` => `node-12` and `nodenext` => `node-next`

> `-` has weird command line parsing implications, so we should probably avoid including it in any options.

Most existing options currently use camelCase (e.g. `--watchFile fixedPollingInterval`), the only exception I can see is `--jsx react-{native,jsx,jsxdev}`.

With that in mind, `nodeNext` would make sense, but from what I can tell the `--module`/`--moduleResolution` flags already don't care about capitalization so this should already work.

File: src/compiler/types.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (ewlsh) [reply src/compiler/types.ts:6141]: As an example/datapoint, GJS ([JavaScript for GNOME](https://gitlab.gnome.org/GNOME/gjs/)) supports ESM so our projects would likely use the ES2015-ESNext module settings.

We support either a root-level command line flag (`-m`/`--module`), a root-level API (`gjs_evaluate_module`), or treat all dynamic imports (`import('./x.js')`) as ESM (regardless of extension).

**Comment by bmeck** [comment]
Does this deal w/ "default" exports per my concern in https://github.com/microsoft/TypeScript/issues/41139#issuecomment-810353181 ?

**Comment by Jack-Works** [comment]
It seems like I must have `type: module` in `package.json` otherwise `module: nodenext` does not work... Why?

File: src/compiler/utilities.ts [hunk 9]
```diff
@@ -6098,8 +6102,11 @@ namespace ts {
         return scriptKind === ScriptKind.TSX || scriptKind === ScriptKind.JSX || scriptKind === ScriptKind.JS || scriptKind === ScriptKind.JSON ? LanguageVariant.JSX : LanguageVariant.Standard;
     }
 
-    export function getEmitScriptTarget(compilerOptions: CompilerOptions) {
-        return compilerOptions.target || ScriptTarget.ES3;
+    export function getEmitScriptTarget(compilerOptions: {module?: CompilerOptions["module"], target?: CompilerOptions["target"]}) {
+        return compilerOptions.target ||
+            (compilerOptions.module === ModuleKind.Node12 && ScriptTarget.ES2020) ||
+            (compilerOptions.module === ModuleKind.NodeNext && ScriptTarget.ESNext) ||
+            ScriptTarget.ES3;
```
Comment (DanielRosenwasser) [inline src/compiler/utilities.ts:6109]: I think I was the one who recommended `module` auto-implying `target`, but it does make me feel a little bit anxious. The fact that the two can affect each other feels weirdly circular.

File: src/services/transpile.ts [hunk 10]
```diff
@@ -48,7 +48,7 @@ namespace ts {
 
         // if jsx is specified then treat file as .tsx
         const inputFileName = transpileOptions.fileName || (transpileOptions.compilerOptions && transpileOptions.compilerOptions.jsx ? "module.tsx" : "module.ts");
-        const sourceFile = createSourceFile(inputFileName, input, options.target!); // TODO: GH#18217
+        const sourceFile = createSourceFile(inputFileName, input, getEmitScriptTarget(options));
```
Comment (DanielRosenwasser) [inline src/services/transpile.ts:51]: Okay, so question about transpile - usually, we look at the `package.json` to figure out if a module needs to be ESM or CJS. How do you control that with `transpileModule`?

File: src/compiler/transformers/module/esnextAnd2015.ts [hunk 11]
```diff
@@ -55,8 +69,10 @@ namespace ts {
         function visitor(node: Node): VisitResult<Node> {
             switch (node.kind) {
                 case SyntaxKind.ImportEqualsDeclaration:
-                    // Elide `import=` as it is not legal with --module ES6
-                    return undefined;
+                    // Though an error in es2020 modules, in node-flavor es2020 modules, we can helpfully transform this to a synthetic `require` call
+                    // To give easy access to a synchronous `require` in node-flavor esm. We do the transform even in scenarios where we error, but `import.meta.url`
+                    // is available, just because the output is reasonable for a node-like runtime.
```
Comment (DanielRosenwasser) [inline src/compiler/transformers/module/esnextAnd2015.ts:74]:
```suggestion
                    // to give easy access to a synchronous `require` in node-flavor esm. We do the transform even in scenarios where we error, but `import.meta.url`
                    // is available, just because the output is reasonable for a node-like runtime.
```

I don't understand the second sentence - how is `import.meta.url` is relevant?

**Comment by Jack-Works** [comment]
> package.json

```json
{
  "type": "module",
  "dependencies": {
    "typescript": "npm:@typescript-deploys/pr-build@4.5.0-pr-45884-2"
  }
}
```

> tsconfig.json
```json
{
    "compilerOptions": {
        "outDir": "./dist",
        "module": "node12",
        "target": "esnext"
    },
    "include": ["./src"]
}

```

> src/index.mts
```ts
import { util } from './deps.mts'
export function main() {
    util()
}
```

> src/deps.mts
```ts
export function util() {}
```

It reports:

> error TS2691: An import path cannot end with a '.mts' extension. Consider importing './deps.js' instead.

But when I changed to `.deps.js`, it errors with

> error TS2307: Cannot find module './deps.js' or its corresponding type declarations.


I changed to `./deps.mjs` it finally works

**Comment by Jack-Works** [comment]
The auto-import also writes `import { util } from "./deps.mts";` instead of `.mjs`

**Comment by weswigham** [comment]
Mmm, that diagnostic message needs to be updated - `.mts` outputs as `.mjs` and `.cts` as `.cjs` - right now it just always writes `.js` (rather than fetching the emitted filename).

**Comment by ctjlewis** [comment]
Excellent work on this PR and mjs/cjs support, Wes. (We have been discussing full ES specifier support in #42151.)

File: src/compiler/diagnosticMessages.json [hunk 12]
```diff
@@ -1389,6 +1389,15 @@
         "code": 1449
     },
 
+    "The 'import.meta' meta-property is not allowed in files which will build into CommonJS output.": {
+        "category": "Error",
+        "code": 1470
+    },
+    "Module '{0}' cannot be imported using this construct. The specifier only resolves to an es module, which cannot be imported synchronously. Use dynamic import instead.": {
```
Comment (andrewbranch) [inline src/compiler/diagnosticMessages.json:1396]: Nit:

```suggestion
    "Module '{0}' cannot be imported using this construct. The specifier only resolves to an ES module, which cannot be imported synchronously. Use a dynamic import instead.": {
```

File: src/services/transpile.ts (hunk 10 above)
Comment (andrewbranch) [reply src/services/transpile.ts:51]: I’m guessing we’d either need an additional for `transpileModule` or the caller can spoof a `.mts` or `.cts` extension since the input file name need not represent a real file on disk. Is that the idea @weswigham?

**Comment by weswigham** [comment]
Merged in #45884

**Comment by mattbishop** [comment]
Are we going to be able to import local modules with .ts filenames, like Deno does?

https://deno.land/manual@v1.14.2/examples/import_export

Basically:
```typescript
import {something} from "../some/things.ts"
```

Naturally the compiler would convert this to `things.js`.

**Comment by orta** [comment]
We're not for node-esm, no. See https://github.com/microsoft/TypeScript/issues/15479#issuecomment-300240856, https://github.com/microsoft/TypeScript/issues/16577#issuecomment-754941937, https://github.com/microsoft/TypeScript/issues/26722#issuecomment-580975983, https://github.com/microsoft/TypeScript/issues/33588.

Though I believe we're pretty open for `module: deno` in the future which I'd expect would have this behavior (because the runtime supports the identifiers being passed in)

**Comment by mattbishop** [comment]
I'm probably naive but wouldn't the result be two kinds of typescript? One for Deno that uses ".ts" module references and one for node and browsers. If you create a module for node, you won't be able to import it to Deno as TS code, you would have to rely on the compiler and .d.ts file. The idea of TS-only modules suffers. 

I suppose this is a Deno issue too. They should probably support .js imports. Aesthetically though, it looks nicer to import .ts files over .js. It means you have to make that mental connection to the fact the compiler generated JavaScript.

**Comment by Jack-Works** [comment]
> you would have to rely on the compiler and .d.ts file

even that won't be possible I guess. Referencing other `.d.ts` file might have different module resolution methods. e.g. some platform requires extensions and some platform requires different ones.

**Comment by Pacjonek** [comment]
Why couldn't Typescript convert file extensions from js to ts during compilation? Importing js files in Typescript projects where you work on ts files looks very strange, very aesthetically unpleasant

**Comment by Pacjonek** [comment]
On the subject of imports, why couldn't Typescript convert file extensions from *.js to *.ts during compilation? So that a programmer who works in a Typescript project imports files with the *.ts extension. 
Importing *.js files in Typescript projects where you work on *.ts files looks... very strange, very aesthetically unpleasant

**Comment by simllll** [comment]
I was testing this new feature in the new beta 4.5, but my await imports are still transformed to old requires.. what I'm missing here?

my tsconfig.json:
```json
{
	"compilerOptions": {
	  "module": "node12",
	  "esModuleInterop": true,
	}
}
```


my .ts file:

```ts
import 'reflect-metadata';

const test = async () => {
	const { MongooseConnection } = await import('@hokify/db-connection');
	const { AgendaScheduler } = await import('@hokify/scheduler');

	// test
};

test()
	.then(() => {
		console.log('sync done');
	})
	.catch((err: any) => {
		console.log('err', err.response?.data?.errors || err);
	});
```

running `tsc test.ts` produces:
```js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
exports.__esModule = true;
require("reflect-metadata");
var test = function () { return __awaiter(void 0, void 0, void 0, function () {
    var MongooseConnection, AgendaScheduler;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0: return [4 /*yield*/, Promise.resolve().then(function () { return require('@hokify/db-connection'); })];
            case 1:
                MongooseConnection = (_a.sent()).MongooseConnection;
                return [4 /*yield*/, Promise.resolve().then(function () { return require('@hokify/scheduler'); })];
            case 2:
                AgendaScheduler = (_a.sent()).AgendaScheduler;
                return [2 /*return*/];
        }
    });
}); };
test()
    .then(function () {
    console.log('sync done');
})["catch"](function (err) {
    var _a, _b;
    console.log('err', ((_b = (_a = err.response) === null || _a === void 0 ? void 0 : _a.data) === null || _b === void 0 ? void 0 : _b.errors) || err);
});

```

, which does not work, because both improts are esm modules, and therefore running this code snippet results into err ```
Error [ERR_REQUIRE_ESM]: require() of ES Module /home/simon/Dev/hokify/hokify-server/packages/db-connection/dist/index.js from /home/simon/Dev/hokify/hokify-server/packages/asana/__scripts__/createTicket.js not supported.
Instead change the require of index.js in /home/simon/Dev/hokify/hokify-server/packages/asana/__scripts__/createTicket.js to a dynamic import() which is available in all CommonJS modules.
    at /home/simon/Dev/hokify/hokify-server/packages/asana/__scripts__/createTicket.js:44:86 {
  code: 'ERR_REQUIRE_ESM'

```

**Comment by simllll** [comment]
I found the reason for the 2nd issue, I still have some troubles to get things running, but I guess this is related to something else now. Anyways, settings "moduleResolution": "node" causes this issue. I somehow extended a tsconfig file in the root, that had this flag set. Removing it, solves everything. Now I'm wondering though if the implcity set moduleResolution to node12 (like it's stated in the PR) is not working correclty?

**Comment by danrivett** [comment]
Thanks @ctjlewis. After my comment I saw @orta's twitter post and how many people have already questioned this, so I don't want to beat it to death.

For the record, I don't disagree about not wanting to change runtime behaviour and how that is set in stone, I just disagree that it is changing runtime behaviour.

I don't see that converting non-suffixed imports into `*.js` suffixed imports as part of the transpile step would be considered a runtime change in behaviour. To me that is a syntax change, with no runtime behaviour change.

Reading about this more it seems this discussion has become quite entrenched so I will leave it at this comment as I think I've articulated it as best I can why I feel this is the wrong conclusion.

**Comment by ctjlewis** [comment]
IMO @danrivett, you're absolutely correct, rewriting imports is compile-time behavior that is *not* a runtime change—in fact, for this resolution algorithm, you are having to make the compile-time changes by hand, in your source code, rather than letting the TS compiler do it for you.

I have personally said enough on this to digress now also, as maintainers have made it clear they feel like it's a waste of their time—but it seems clear that developers would prefer to not rewrite thousands of imports by hand.

**Comment by benjie** [comment]
Hello, your beta announcements are asking for feedback on this feature. First of all - I'm super excited about this! Great work!

One thing I was really looking forward to in Node.js exports/ESM support was having example files that import from the library itself - this way we can just say `import { foo } from "my-module"` rather than `import { foo } from ".." /* TODO: replace this with "my-module" */` - makes it much easier for users to copy/paste examples whilst keeping the examples co-located inside the package.

Node.js supports this "self-referencing" behaviour; however TypeScript doesn't seem to understand how to import this in `module:"nodenext"` mode - a `could not find a declaration file for module 'my-module'` error is raised.

I've prepared a minimal reproduction here with `js` folder working using Node.js` and analogous `ts` folder which is not working with TypeScript (won't compile):

https://github.com/benjie/typescript-self-referencing-issue

TSConfig:

```json
{
  "compilerOptions": {
    "lib": ["es2021"],
    "module": "nodenext",
    "target": "es2021",

    "strict": true,

    "rootDir": "src",
    "declaration": true,
    "declarationDir": "dist",
    "outDir": "dist"
  },
  "include": ["src"],
  "exclude": ["dist"]
}
```

Thanks for your hard work on TypeScript :heart:

**Comment by andrewbranch** [comment]
@benjie thanks for the feedback! It works after you do a build once and `dist/index.d.ts` is present. I think this is a bug—it seems like we need to map back from `outDir` to the root. I’ve copied your report to #46762.

**Comment by benjie** [comment]
Thanks @andrewbranch! When I did a second build afterwards it told me I was trying to overwrite source files; I tried adding exclude but that didn’t solve it.

**Comment by arpowers** [comment]
@orta on the topic of extensions, there needs to be more clarification about how to handle things during development. Specific  questions:  
- The `import/no unresolved` eslint rule complains about the file not existing, recommend just disabling that rule?
- Due to this change/standardization of convention regarding file extensions, what is a good general way of "always" handling extensions across a large TS project? (working across both browser and node)

**Comment by loynoir** [comment]
sugar `import = require` to `require-or-createRequire` and allow `try-catch`?

**Comment by nayeemrmn** [comment]
Hello, one of the features gated by `"moduleResolution": "nodenext"` is the per-export types mentioned at https://devblogs.microsoft.com/typescript/announcing-typescript-4-5-beta/#packagejson-exports-imports-and-self-referencing.

Basically if a dependency has a `package.json` like:
```json
{
  "exports": {
    "./subpath": {
      "import": "./esm/subpath/index.js",
      "require": "./commonjs/subpath/index.js",
      "types": "./types/subpath/index.d.ts"
    }
  }
}
```
the `"types"` field above will not be looked up by a _user_ of that dependency unless _they_ have `"moduleResolution": "nodenext"` in their tsconfig.

Would the team be open to extending this feature to `"moduleResolution": "node"`? It's difficult that a library opting into the `exports["subpath"].types` feature extends a config requirement to its users. Sure that's the point of `"moduleResolution"` but I noticed that the other features gated by `"moduleResolution": "nodenext"` are more in the spirit of organising your own code. I think this would be very non-breaking in practice.

reverse ref https://github.com/denoland/dnt/issues/205

**Comment by weswigham** [comment]
No. Old node (ie, pre 12), doesn't support exports. So it won't be getting support for resolution features the matching runtime doesn't have. :)

See #50153 for what we're actually thinking about doing - we know stuff like deno wants more flexible resolution, but we're not going to reduce safety for people using other platforms to achieve it.

**Comment by nayeemrmn** [comment]
> See #50153 for what we're actually thinking about doing - we know stuff like deno wants more flexible resolution, but we're not going to reduce safety for people using other platforms to achieve it.

No worries, some conflation there as it's not a Deno concern at all but just a quirk of dnt's current exports generation (for node) that made me notice this. Fair points otherwise. To pivot slightly, the default `"moduleResolution"` value will eventually be one that supports exports, right?

> (Also, as an aside, that example export map, while vaguely functional, is incomplete - it shares types between two formats of entry points, which will create subtle (or not so subtle if the types are esm format) bugs on cross format-interop checking. Each of the `require` and `import` conditions should have it's _own_ nested types entry.

I got that from https://devblogs.microsoft.com/typescript/announcing-typescript-4-5-beta/#packagejson-exports-imports-and-self-referencing. Is there a syntax for giving different types to each condition?

**Comment by weswigham** [comment]
> the default "moduleResolution" value will eventually be one that supports exports, right?

The `tsc --init` default will probably be updated soonish, maybe. We'll probably discuss it after we handle that issue I linked. The compiler default we may change, but likely on a much longer timescale, since it's a breaking change.

**Comment by weswigham** [comment]
> I got that from

@DanielRosenwasser can we get that example fixed? I thought we did ages ago - it should be
```json
// package.json
{
    "name": "my-package",
    "type": "module",
    "exports": {
        ".": {
            // Entry-point for `import "my-package"` in ESM
            "import": 
                // Entry-point for TypeScript resolution
                "types": "./types-esm/index.d.ts"
                 // Runtime entry-point
                "default": "./esm/index.js"
            },

            // Entry-point for `require("my-package") in CJS
            "require": {
                // Entry-point for TypeScript resolution
                "types": "./types-cjs/index.d.cts"
                 // Runtime entry-point
                "default": "./commonjs/index.cjs"
            },
        },
    },

    // CJS fall-back for older versions of Node.js
    "main": "./commonjs/index.cjs",

    // Fall-back for older versions of TypeScript (will usually be treated as cjs by older TS)
    "types": "./types-esm/index.d.ts"
}
```

**Comment by nayeemrmn** [comment]
Also I got to the 4.5 beta release post from https://www.typescriptlang.org/tsconfig#moduleResolution, that should be upgraded to the 4.7 release post.

**Comment by lgarron** [comment]
> The compiler default we may change, but likely on a much longer timescale, since it's a breaking change.

I’m really looking forward to this point, since it will allow me as a library author to simplify and deduplicate some “gotcha” workarounds. Is there an appropriate place to track this/would it be alright to file a new issue for that?

**Comment by weswigham** [comment]
I don't think we have an issue tracking it, it's just something we're keeping at the back of our minds.

**Comment by thw0rted** [comment]
> Rule of thumb: one js file, one .d.ts file

@weswigham  what does that mean for community-maintained packages in DefinitelyTyped?  Say we're typing a CJS library that starts shipping dual-package using `exports`, but continues to avoid shipping in-package type defs.  Should the corresponding `@types` package ship an `exports` field with separate .d.ts files under `exports.import.types` and `exports.require.types`?
